### PR TITLE
codewhisperer: update zipUtil for full project scans

### PR DESCRIPTION
**Pull request by ctlai95**
## Problem

Add support for full project scans

## Solution

- Decouple scans from dependency graph logic
- Refactor startSecurityScan method to support Project scan type
- Add `ignore` package to exclude ignored files from artifact
- Include all Java `.class` files in the artifact

## License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by ashishrp-aws — APPROVED** [review]

File: packages/core/src/codewhisperer/util/zipUtil.ts [hunk 1]
```diff
@@ -5,11 +5,13 @@
 import admZip from 'adm-zip'
 import * as vscode from 'vscode'
 import path from 'path'
-import { tempDirPath } from '../../shared/filesystemUtilities'
+import { readFileAsString, tempDirPath } from '../../shared/filesystemUtilities'
 import { getLogger } from '../../shared/logger'
 import * as CodeWhispererConstants from '../models/constants'
+import { existsSync, readdirSync, statSync } from 'fs'
```
Comment (nkomonen-amazon) [inline packages/core/src/codewhisperer/util/zipUtil.ts:11]: Going forward we are avoiding the use of `fs` and `fs-extra` for the extension to be compatible in the browser.

Please look at `FileSystemCommon` and start using that instead. It should have what you need.

File: packages/core/src/codewhisperer/util/zipUtil.ts [hunk 2]
```diff
@@ -61,8 +72,61 @@ export class ZipUtil {
...
+            return size > this.getProjectScanPayloadSizeLimitInBytes()
+        }
+    }
+
+    public willReachSizeLimit(current: number, adding: number): boolean {
+        const willReachLimit = current + adding > this.getProjectScanPayloadSizeLimitInBytes()
+        return willReachLimit
+    }
+
+    protected async traverseDir(dirPath: string, ignore?: Ignore) {
```
Comment (justinmk3) [inline packages/core/src/codewhisperer/util/zipUtil.ts:92]: This is doing a lot more than "traverse dir". What is it actually doing? The docstring should say so. Or use a more meaningfull name, such as `collectFiles`
```suggestion
    /** Recursively add files to ... and compute stats ... */
    protected async collectFiles(dirPath: string, ignore?: Ignore) {
```

Please, make this a habit for all non-obvious code in the future.

File: packages/core/src/test/codewhisperer/util/zipUtil.test.ts [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+import { join } from 'path'
+import { getTestWorkspaceFolder } from '../../../testInteg/integrationTestsUtilities'
+import { ZipUtil } from '../../../codewhisperer/util/zipUtil'
+import { SecurityScanType } from '../../../codewhisperer/models/constants'
+import * as CodeWhispererConstants from '../../../codewhisperer/models/constants'
+import { ToolkitError } from '../../../shared/errors'
+import fs from 'fs'
+import { fsCommon } from '../../../srcShared/fs'
+
+describe('zipUtil', function () {
```
Comment (justinmk3) [inline packages/core/src/test/codewhisperer/util/zipUtil.test.ts:18]: We are building a lot of scar tissue around adm-zip, but still haven't implemented a "streaming zip" approach. 

So we're building tech debt that gets us even further from our desired destination.

Comment (ctlai95) [reply packages/core/src/test/codewhisperer/util/zipUtil.test.ts:18]: I have the streaming zip approach ready locally, will raise another PR after this one

File: package-lock.json [hunk 4]
```diff
@@ -11662,9 +11662,9 @@
             "integrity": "sha512-4vf7I2LYV/HaWerSo3XmlMkp5eZ83i+/CDluXi/IGTs/O1sejBNhTtnxzmRZfvOUqj7lZjqHkeTvpgSFDlWZTg=="
         },
         "node_modules/ignore": {
-            "version": "5.2.4",
-            "resolved": "https://registry.npmjs.org/ignore/-/ignore-5.2.4.tgz",
-            "integrity": "sha512-MAb38BcSbH0eHNBxn7ql2NH/kX33OkB3lZ1BNdh7ENeRChHTYsTvWrMubiIAMNS2llXEEgZ1MUOBtXChP3kaFQ==",
+            "version": "5.3.1",
+            "resolved": "https://registry.npmjs.org/ignore/-/ignore-5.3.1.tgz",
+            "integrity": "sha512-5Fytz/IraMjqpwfd34ke28PTVMjZjJG2MPn5t7OE4eUCUNf8BAa7b5WUS9/Qvr6mwOQS7Mk6vdsMno5he+T8Xw==",
```
Comment (justinmk3) [inline package-lock.json:11667]: We can revert this now, right?

Comment (ctlai95) [reply package-lock.json:11667]: Yes let me revert it

File: packages/core/src/codewhisperer/util/zipUtil.ts [hunk 5]
```diff
@@ -8,8 +8,9 @@ import path from 'path'
 import { tempDirPath } from '../../shared/filesystemUtilities'
 import { getLogger } from '../../shared/logger'
 import * as CodeWhispererConstants from '../models/constants'
+import { ToolkitError } from '../../shared/errors'
 import { fsCommon } from '../../srcShared/fs'
-import { statSync } from 'fs'
+import { collectFiles } from '../../amazonqFeatureDev/util/files'
```
Comment (justinmk3) [inline packages/core/src/codewhisperer/util/zipUtil.ts:13]: Nice! Glad this was able to be re-used. That really gives us leverage.

In a separate PR, consider moving `collectFiles` somewhere out of "featureDev" (perhaps to the existing `workspaceUtils.ts` module)

File: packages/core/src/codewhisperer/util/zipUtil.ts [hunk 6]
```diff
@@ -76,12 +99,48 @@ export class ZipUtil {
...
+            if (this.isJavaClassFile(file.fileUri)) {
+                this._pickedBuildFiles.add(file.fileUri.fsPath)
+                this._totalBuildSize += fileSize
+            } else {
+                if (
+                    this.reachSizeLimit(this._totalSize, CodeWhispererConstants.SecurityScanType.Project) ||
+                    this.willReachSizeLimit(this._totalSize, fileSize)
+                ) {
+                    getLogger().error(`Payload size limit reached`)
+                    throw new ToolkitError('Payload size limit reached.')
```
Comment (justinmk3) [inline packages/core/src/codewhisperer/util/zipUtil.ts:137]: NB: errors are automatically logged
```suggestion
                    throw new ToolkitError('Payload size limit reached.')
```

**Review by justinmk3 — APPROVED** [review]
Huge improvement!

File: packages/core/src/codewhisperer/util/zipUtil.ts (hunk 5 above)
Comment (ctlai95) [reply packages/core/src/codewhisperer/util/zipUtil.ts:13]: Thanks! I will work on that separately

**Comment by ctlai95** [comment]
Looking into test failures

File: packages/core/src/test/codewhisperer/commands/startSecurityScan.test.ts [hunk 7]
```diff
@@ -316,7 +316,5 @@ describe('startSecurityScan', function () {
         )
         assert.ok(commandSpy.neverCalledWith('workbench.action.problems.focus'))
         assert.ok(securityScanRenderSpy.notCalled)
-        const warnings = getTestWindow().shownMessages.filter(m => m.severity === SeverityLevel.Warning)
-        assert.strictEqual(warnings.length, 0)
```
Comment (justinmk3) [inline packages/core/src/test/codewhisperer/commands/startSecurityScan.test.ts:320]: If this assertion is "flaky" does that mean an unexpected message is being shown?

**Pull request merged**